cmd/vsphere-xcopy-volume-populator/internal/pure/flashArray.go

```diff
@@ -0,0 +1,192 @@
+	g, err := f.client.Hostgroups.GetHostgroup(initiatorGroup, nil)
+	if err != nil {
+		if strings.HasPrefix(err.Error(), "Response code: 400") {
+			klog.Infof("not found - now created it group named %v", err)
+			if g, err = f.client.Hostgroups.CreateHostgroup(initiatorGroup, nil); err != nil {
+				return nil, fmt.Errorf("failed creating host group with name %v: %w", initiatorGroup, err)
+			}
+			return populator.MappingContext{}, nil
+
+		} else {
+			klog.Infof("error getting host group named %v", initiatorGroup)
+
+			return nil, err
+		}
+	}
+
+	hosts, err := f.client.Hosts.ListHosts(nil)
+	if err != nil {
+		return nil, err
+	}
+	for _, h := range hosts {
+		for _, iqn := range h.Iqn {
+			if slices.Contains(clonnerIqn, iqn) {
+				klog.Infof("adding host to group %v", h.Name)
+				g.Hosts = append(g.Hosts, h.Name)
+			}
+		}
+		for _, wwn := range h.Wwn {
+			if slices.Contains(clonnerIqn, wwn) {
+				klog.Infof("adding host to group %v", h.Name)
+				g.Hosts = append(g.Hosts, h.Name)
+			}
+		}
+	}
+	klog.Infof("hosts to to set in group %v", g.Hosts)
+	g, err = f.client.Hostgroups.SetHostgroup(g.Name, map[string][]string{"hostlist": g.Hosts})
+	if err != nil {
+		return populator.MappingContext{}, err
+	}
+
+	return populator.MappingContext{"hosts": g.Hosts}, nil
+}
+
+// Map is responsible to mapping an initiator group to a populator.LUN
+func (f *FlashArrayClonner) Map(
+	initatorGroup string,
+	targetLUN populator.LUN,
+	context populator.MappingContext) (populator.LUN, error) {
+	hosts, ok := context["hosts"]
+	if ok {
+		hs, ok := hosts.([]string)
+		if ok && len(hs) > 0 {
+			for _, host := range hs {
+				klog.Infof("connecting host %s to volume %s", host, targetLUN.Name)
+				_, err := f.client.Hosts.ConnectHost(host, targetLUN.Name, nil)
+				if err != nil {
+					if strings.Contains(err.Error(), "Connection already exists.") {
+						continue
+					}
+					return populator.LUN{}, err
+				}
+
+			}
+		}
+	}
+
+	if true {
+		return targetLUN, nil
+	}
+
+	g, err := f.client.Hostgroups.GetHostgroup(initatorGroup, nil)
+	if err != nil {
+		return populator.LUN{}, err
+	}
+	connectedVolume, err := f.client.Hostgroups.ConnectHostgroup(g.Name, targetLUN.Name, nil)
+	if err != nil {
+		return populator.LUN{}, nil
+	}
+	klog.Infof("target LUN %v connected to volume %+v", targetLUN, connectedVolume)
+	return targetLUN, nil
+}
+
+// UnMap is responsible to unmapping an initiator group from a populator.LUN
+func (f *FlashArrayClonner) UnMap(initatorGroup string, targetLUN populator.LUN, context populator.MappingContext) error {
+	hosts, ok := context["hosts"]
+	if ok {
+		hs, ok := hosts.([]string)
+		if ok && len(hs) > 0 {
+			for _, host := range hs {
+				klog.Infof("disconnecting host %s from volume %s", host, targetLUN.Name)
+				_, err := f.client.Hosts.DisconnectHost(host, targetLUN.Name)
+				if err != nil {
+					return err
+				}
+
+			}
+		}
+	}
+	return nil
```

Comment:
if you connected lun to hgroup directly, then just disconnect disconnect hgroup. no need to iterate over hosts and manipulate connections, which defeats the purpose of hgroups.